server/snapshot.go

```diff
@@ -22,56 +34,208 @@ func newChannelSnapshot(c *channel) raft.FSMSnapshot {
 // Persist should dump all necessary state to the WriteCloser 'sink',
 // and call sink.Close() when finished or call sink.Cancel() on error.
 func (c *channelSnapshot) Persist(sink raft.SnapshotSink) (err error) {
-	// TODO: this is very expensive and might repeatedly fail if messages are
-	// constantly being truncated. Is there a way we can optimize this, e.g.
-	// handling Restore() out-of-band from Raft?
-
-	// TODO: this needs to also snapshot subscription state.
-
 	defer func() {
 		if err != nil {
 			sink.Cancel()
 		}
 	}()
 
-	var (
-		first uint64
-		last  uint64
-		msg   *pb.MsgProto
-		data  []byte
-	)
-	first, last, err = c.channel.store.Msgs.FirstAndLastSequence()
+	if err := c.snapshotMessages(sink); err != nil {
+		return err
+	}
+
+	if err := c.snapshotSubscriptions(sink); err != nil {
+		return err
+	}
+
+	if err := c.snapshotAcks(sink); err != nil {
+		return err
+	}
+
+	return sink.Close()
+}
+
+// Release is a no-op.
+func (c *channelSnapshot) Release() {}
+
+func (c *channelSnapshot) snapshotMessages(sink raft.SnapshotSink) error {
+	// TODO: this is very expensive and might repeatedly fail if messages are
+	// constantly being truncated. Is there a way we can optimize this, e.g.
+	// handling Restore() out-of-band from Raft?
+
+	first, last, err := c.store.Msgs.FirstAndLastSequence()
 	if err != nil {
 		return err
 	}
 
-	buf := make([]byte, 4)
+	var (
+		buf              [4]byte
+		batch            = batchPool.Get().(*spb.Batch)
+		writeMsgFragment = func(b *spb.Batch) error {
+			fragment := newFragment()
+			fragment.FragmentType = spb.RaftSnapshotFragment_Messages
+			fragment.MessageBatch = b
+			return writeFragment(sink, fragment, buf)
+		}
+	)
+	batch.Messages = make([]*pb.MsgProto, 0, snapshotBatchSize)
+
 	for seq := first; seq <= last; seq++ {
-		msg, err = c.channel.store.Msgs.Lookup(seq)
+		msg, err := c.store.Msgs.Lookup(seq)
 		if err != nil {
-			return
+			batchPool.Put(batch)
+			return err
 		}
 		// If msg is nil, channel truncation has occurred while snapshotting.
 		if msg == nil {
 			// Channel truncation has occurred while snapshotting.
-			return fmt.Errorf("channel %q was truncated while snapshotting", c.channel.name)
+			batchPool.Put(batch)
+			return fmt.Errorf("channel %q was truncated while snapshotting", c.name)
 		}
-		data, err = msg.Marshal()
-		if err != nil {
-			return
+
+		// Previous batch is full, ship it.
+		if len(batch.Messages) == snapshotBatchSize {
```

Comment:
Would we ever want to add a timer in here? Is it possible that it could take longer to read/send a batch than a client timeout?

Reply:
Can you clarify what you mean? When a snapshot is taken, it's streamed out to disk. After the snapshot is saved, the Raft log is truncated. Then later, if a follower needs to catch up, the leader streams the snapshot to the follower (if it doesn't have the actual log entries). The batching is just an optimization since we would otherwise spend a lot of time serializing/deserializing each individual message.